geo/geogfn: implement ST_Azimuth({geometry,geometry})

hueypark opened the pull request:
Fixes #48887

Release note (sql change): This PR implement adds the ST_Azimuth({geometry,geometry})

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50188)
<!-- Reviewable:end -->

otan commented on pkg/geo/geomfn/azimuth_test.go:
```diff
@@ -0,0 +1,71 @@
+	zero            = 0.0
+	aQuarterPi      = 0.7853981633974483
+	towQuartersPi   = 1.5707963267948966
+	threeQuartersPi = 2.356194490192344
```
do you mind moving these vars into the `TestAzimuth` function? (unless we're planning to re-use them)

hueypark replied:
Done.

otan commented on pkg/geo/geomfn/azimuth.go:
```diff
@@ -0,0 +1,29 @@
+		return nil
+	}
+
+	azimuth := math.Mod(2*math.Pi+math.Pi/2-math.Atan2(b.Y()-a.Y(), b.X()-a.X()), 2*math.Pi)
```
do you mind adding a comment of why we're adding 2*math.Pi, i.e. we always want a positive number?
some commentary on that math would also be nice :)

hueypark replied:
Done.

otan reviewed (commented):
quick comment request :)

otan commented on pkg/geo/geomfn/azimuth.go:
```diff
@@ -0,0 +1,29 @@
+// Azimuth returns the azimuth in radians of the segment defined by the given point geometries.
+// The azimuth is angle is referenced from north, and is positive clockwise.
+// North = 0; East = π/2; South = π; West = 3π/2.
+func Azimuth(a *geom.Point, b *geom.Point) *float64 {
```
actually, do you mind moving the logic in geo_builtins such that Azimuth only takes in `(a *geo.Geometry, b *geo.Geometry)`, to be consistent with the rest of the function signatures in ehre?

hueypark replied:
Done.

hueypark committed:
geo/geogfn: implement ST_Azimuth({geometry,geometry})

Fixes #48887

Release note (sql change): This PR implement adds the
`ST_Azimuth({geometry,geometry})` builtin.

hueypark commented:
Thank you for the review!

otan reviewed (approved):
thanks for your contribution!

otan commented:
bors r+
